File: frontend/src/api/categories.js
// Category API calls - Public viewing and Admin management

import axios from 'axios';

const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8001';

/**
 * Create an axios instance with authentication token.
 *
 * @param {string} token - JWT access token
 * @returns {AxiosInstance} Authenticated axios instance
 */
const getApiWithToken = (token) => {
  return axios.create({
    baseURL: API_URL,
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
  });
};

// Public endpoints - No authentication required

/**
 * Fetch all categories with pagination.
 *
 * @returns {Promise<CategoryResponse[]>} List of categories
 */
export const getAllCategories = async () => {
  const response = await axios.get(`${API_URL}/api/categories`);
  return response.data;
};

/**
 * Fetch a single category by ID.
 *
 * @param {number} id - Category ID
 * @returns {Promise<CategoryResponse>} Category data
 */
export const getCategory = async (id) => {
  const response = await axios.get(`${API_URL}/api/categories/${id}`);
  return response.data;
};

// Admin endpoints - Authentication required

/**
 * Create a new category.
 *
 * @param {CategoryCreate} categoryData - Category data (name, slug, description)
 * @param {string} token - JWT access token
 * @returns {Promise<CategoryResponse>} Created category
 */
export const createCategory = async (categoryData, token) => {
  const api = getApiWithToken(token);
  const response = await api.post('/api/admin/categories', categoryData);
  return response.data;
};

/**
 * Update an existing category.
 *
 * @param {number} id - Category ID to update
 * @param {CategoryUpdate} categoryData - Updated category data
 * @param {string} token - JWT access token
 * @returns {Promise<CategoryResponse>} Updated category
 */
export const updateCategory = async (id, categoryData, token) => {
  const api = getApiWithToken(token);
  const response = await api.put(`/api/admin/categories/${id}`, categoryData);
  return response.data;
};

/**
 * Delete a category.
 *
 * @param {number} id - Category ID to delete
 * @param {string} token - JWT access token
 * @returns {Promise<{message: string}>} Deletion confirmation
 */
export const deleteCategory = async (id, token) => {
  const api = getApiWithToken(token);
  const response = await api.delete(`/api/admin/categories/${id}`);
  return response.data;
};

File: backend/app/routes/categories.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from ..database import get_db
from .. import schemas, crud, auth

router = APIRouter()


# Public endpoints - No authentication required

@router.get("/categories", response_model=List[schemas.CategoryResponse])
def get_categories(
    skip: int = 0,
    limit: int = 100,
    db: Session = Depends(get_db)
):
    """
    Retrieve all project categories.
    
    - Public endpoint accessible to all visitors
    - Supports pagination via skip and limit parameters
    """
    return crud.get_all_categories(db, skip=skip, limit=limit)


@router.get("/categories/{category_id}", response_model=schemas.CategoryResponse)
def get_category(
    category_id: int,
    db: Session = Depends(get_db)
):
    """
    Retrieve a single category by its ID.
    
    - Public endpoint accessible to all visitors
    - Returns 404 if category does not exist
    """
    category = crud.get_category(db, category_id)
    if not category:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )
    return category


# Admin endpoints - Authentication required

@router.post("/admin/categories", response_model=schemas.CategoryResponse)
def create_category(
    category: schemas.CategoryCreate,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Create a new category.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Ensures slug uniqueness before creation
    """
    existing = crud.get_category_by_slug(db, category.slug)
    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Category with this slug already exists"
        )
    return crud.create_category(db, category)


@router.put("/admin/categories/{category_id}", response_model=schemas.CategoryResponse)
def update_category(
    category_id: int,
    category_update: schemas.CategoryUpdate,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Update an existing category.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Returns 404 if category does not exist
    """
    category = crud.update_category(db, category_id, category_update)
    if not category:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )
    return category


@router.delete("/admin/categories/{category_id}")
def delete_category(
    category_id: int,
    db: Session = Depends(get_db),
    current_user = Depends(auth.get_current_user)
):
    """
    Delete a category.
    
    - Admin only endpoint
    - Requires valid authentication token
    - Returns 404 if category does not exist
    """
    if not crud.delete_category(db, category_id):
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Category not found"
        )
    return {"message": "Category deleted successfully"}